xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc

```diff
@@ -152,70 +156,88 @@ bool IsAlignedSlice(const HloInstruction* slice) {
+    VLOG(5) << "Computation caller is not while, it is "
+            << while_op->ToString();
+    return false;
+  }
+  HloComputation* while_body = while_op->while_body();
+  std::optional<int64_t> tuple_idx = GetLoopInductionVarTupleIdx(while_op);
+  if (!tuple_idx.has_value()) {
+    VLOG(5) << "Induction variable tuple index is nullopt";
     return false;
   }
-  // Check the update operation
-  const HloInstruction* updated_var =
-      offset.parent()->root_instruction()->operand(param_idx);
-  if (!Match(updated_var, m::Add(m::GetTupleElement(m::Parameter(0), param_idx),
-                                 m::ConstantScalar(1)))) {
+  if (while_body->num_parameters() != 1) {
+    VLOG(5) << "Expected only one parameter for while body: "
+            << while_op->ToString();
     return false;
   }
-  // Check that the condition considers this.
-  const HloInstruction* condition_root =
-      while_instr->while_condition()->root_instruction();
-  if (!Match(condition_root,
-             m::Lt(m::GetTupleElement(m::Parameter(0), param_idx),
-                   m::ConstantScalar(trip_count)))) {
+  if (!while_body->parameter_instruction(0)->shape().IsTuple()) {
+    VLOG(5) << "Expected while body param to be a tuple, found "
+            << while_op->ToString();
     return false;
   }
-  // Check init
-  const HloInstruction* init_loop_iter =
-      while_instr->operand(0)->operand(param_idx);
-  if (!Match(init_loop_iter, m::ConstantScalar(0))) {
+  HloInstruction* indvar = hlo_query::GetUniqueGteInstruction(
+      while_body->parameter_instruction(0), *tuple_idx);
+  if (!indvar) {
+    VLOG(5) << "Unable to find unique GTE for while induction variable idx: "
+            << *tuple_idx << ", while op: " << while_op->ToString();
     return false;
   }
+  const HloInstruction* update =
+      while_body->root_instruction()->operand(*tuple_idx);
+  std::unique_ptr<HloModule> update_module = ExtractModule(update);
+  std::unique_ptr<HloModule> value_module = ExtractModule(&value);
+  auto is_dependent_on_indvar_only = [tuple_idx](HloModule* module) {
+    HloComputation* entry = module->entry_computation();
+    if (entry->num_parameters() != 1) {
+      VLOG(5) << "Entry computation has more than one parameter : "
+              << module->ToString();
+      return false;
+    }
+    HloInstruction* param = entry->parameter_instruction(0);
+    // Check that all users of this param are gte
+    if (!absl::c_all_of(
+            param->users(), [tuple_idx](const HloInstruction* user) {
+              return user->opcode() == HloOpcode::kGetTupleElement &&
+                     user->tuple_index() == *tuple_idx;
+            })) {
+      VLOG(5)
+          << "Not all users of the param are gte operations with tuple idx: "
+          << *tuple_idx << ", module: " << module->ToString();
+      return false;
+    }
 
-  return true;
+    // if any instruction is partition-id or replica-id then this is not valid.
```

Comment:
Can you expand on why? My assumption is that this is because it inherently introduces additional, implicit parameters to the computation.

Reply:
Yes, you are correct. The host needs to know this (shouldn't be hard, but trying to support that will be the next step, after this offset work gets done.)